zerver/tests/test_push_notifications.py

```diff
@@ -1490,6 +1495,166 @@ def test_register_push_device_error(self) -> None:
+        with (
+            mock.patch(
+                "zerver.worker.register_push_device_to_bouncer.do_register_remote_push_device",
+                side_effect=InvalidBouncerPublicKeyError,
+            ),
+            self.capture_send_event_calls(expected_num_events=1) as events,
+        ):
+            result = self.client_post("/json/mobile_push/register", payload)
+        self.assert_json_success(result)
+        push_devices = PushDevice.objects.all()
+        self.assert_length(push_devices, 0)
+        self.assertEqual(
+            events[0]["event"],
+            dict(
+                type="push_account",
+                op="remove",
+                push_account_id="2408",
+                reason="Invalid bouncer_public_key",
+            ),
+        )
+
+        # Verify PushRegistrationLivenessTimedOutError
+        with (
+            mock.patch(
+                "zerver.worker.register_push_device_to_bouncer.do_register_remote_push_device",
+                side_effect=PushRegistrationLivenessTimedOutError,
+            ),
+            self.assertLogs(level="ERROR") as m,
+            self.capture_send_event_calls(expected_num_events=1) as events,
+        ):
+            result = self.client_post("/json/mobile_push/register", payload)
+        self.assert_json_success(result)
+        push_devices = PushDevice.objects.all()
+        self.assert_length(push_devices, 0)
+        self.assertEqual(
+            events[0]["event"], dict(type="push_account", op="remove", push_account_id="2408")
+        )
+        self.assertEqual(
+            m.output,
+            [
+                f"ERROR:root:Push device registration for user_id={hamlet.id}, push_account_id=2408 timedout."
+            ],
+        )
+
+        # Verify any other Exception resulting in retry and eventually timedout error.
+        attempt = [0]
+        start_time = now()
+
+        def fake_time() -> int:
+            if attempt[0] == 1:
+                return datetime_to_timestamp(start_time + timedelta(days=2))
+            return datetime_to_timestamp(start_time + timedelta(hours=attempt[0]))
+
+        def raise_exception(*args: Any, **kwargs: Any) -> None:
+            attempt[0] += 1
+            raise Exception("some failure")
+
+        with (
+            mock.patch(
+                "zerver.worker.register_push_device_to_bouncer.do_register_remote_push_device",
+                side_effect=raise_exception,
+            ),
+            mock.patch(
+                "zerver.worker.register_push_device_to_bouncer.time.time", side_effect=fake_time
+            ),
+            mock.patch(
+                "zerver.worker.register_push_device_to_bouncer.time.sleep", return_value=None
+            ),
+            self.assertLogs(level="WARNING"),
+            self.capture_send_event_calls(expected_num_events=1) as events,
+        ):
+            result = self.client_post("/json/mobile_push/register", payload)
```

Comment:
A benefit of my proposal to use normal functions is you don't need to put so much mocking into so many parts of your tests. Thought `time` manipulation in general should be using the `time_machine` module.